ext/bcmath/bcmath.c

```diff
@@ -777,4 +791,938 @@ PHP_FUNCTION(bcscale)
+	}
+	*ret = bc_add(n1, n2, *scale);
+	(*ret)->n_scale = MIN(*scale, (*ret)->n_scale);
+	bc_rm_trailing_zeros(*ret);
+}
+
+static zend_always_inline void bcmath_number_sub_internal(
+	bc_num n1, bc_num n2, bc_num *ret,
+	size_t n1_full_scale, size_t n2_full_scale, size_t *scale, bool auto_scale
+) {
+	if (auto_scale) {
+		*scale = MAX(n1_full_scale, n2_full_scale);
+	}
+	*ret = bc_sub(n1, n2, *scale);
+	(*ret)->n_scale = MIN(*scale, (*ret)->n_scale);
+	bc_rm_trailing_zeros(*ret);
+}
+
+static zend_always_inline void bcmath_number_mul_internal(
+	bc_num n1, bc_num n2, bc_num *ret,
+	size_t n1_full_scale, size_t n2_full_scale, size_t *scale, bool auto_scale
+) {
+	if (auto_scale) {
+		*scale = MIN(n1_full_scale + n2_full_scale, INT_MAX);
+	}
+	*ret = bc_multiply(n1, n2, *scale);
+	(*ret)->n_scale = MIN(*scale, (*ret)->n_scale);
+	bc_rm_trailing_zeros(*ret);
+}
+
+static zend_always_inline zend_result bcmath_number_div_internal(
+	bc_num n1, bc_num n2, bc_num *ret,
+	size_t n1_full_scale, size_t n2_full_scale, size_t *scale, bool auto_scale
+) {
+	if (auto_scale) {
+		*scale = MIN(n1_full_scale + BC_MATH_NUMBER_MAX_EX_SCALE, INT_MAX);
+	}
+	if (!bc_divide(n1, n2, ret, *scale)) {
+		zend_throw_exception_ex(zend_ce_division_by_zero_error, 0, "Division by zero");
+		return FAILURE;
+	}
+	bc_rm_trailing_zeros(*ret);
+	if (auto_scale) {
+		size_t diff = *scale - (*ret)->n_scale;
+		*scale -= diff > BC_MATH_NUMBER_MAX_EX_SCALE ? BC_MATH_NUMBER_MAX_EX_SCALE : diff;
+	}
+	return SUCCESS;
+}
+
+static zend_always_inline zend_result bcmath_number_mod_internal(
+	bc_num n1, bc_num n2, bc_num *ret,
+	size_t n1_full_scale, size_t n2_full_scale, size_t *scale, bool auto_scale
+) {
+	if (auto_scale) {
+		*scale = MAX(n1_full_scale, n2_full_scale);
+	}
+	if (!bc_modulo(n1, n2, ret, *scale)) {
+		zend_throw_exception_ex(zend_ce_division_by_zero_error, 0, "Modulo by zero");
+		return FAILURE;
+	}
+	bc_rm_trailing_zeros(*ret);
+	return SUCCESS;
+}
+
+static zend_always_inline zend_result bcmath_number_pow_internal(
+	bc_num n1, bc_num n2, bc_num *ret,
+	size_t n1_full_scale, size_t n2_full_scale, size_t *scale, bool auto_scale, bool is_op
+) {
+	/* Check the exponent for scale digits and convert to a long. */
+	if (UNEXPECTED(n2->n_scale != 0)) {
+		if (is_op) {
+			zend_value_error("exponent cannot have a fractional part");
+		} else{
+			zend_argument_value_error(1, "exponent cannot have a fractional part");
+		}
+		return FAILURE;
+	}
+	long exponent = bc_num2long(n2);
+
+	bool scale_expand = false;
+	if (auto_scale) {
+		if (exponent > 0) {
+			*scale = n1_full_scale * exponent;
+			if (UNEXPECTED(*scale > INT_MAX || *scale < n1_full_scale)) {
+				*scale = INT_MAX;
+			}
+		} else if (exponent < 0) {
+			*scale = MIN(n1_full_scale + BC_MATH_NUMBER_MAX_EX_SCALE, INT_MAX);
+			scale_expand = true;
+		} else {
+			*scale = 0;
+		}
+	}
+
+	if (UNEXPECTED(exponent == 0 && (n2->n_len > 1 || n2->n_value[0] != 0))) {
+		if (is_op) {
```

Comment:
I don't understand these error messages. "exponent is too large", but it's equal to 0?

Reply:
This is rather confusing, but the existing functions already do this.

When the argument exponent is too large, this variable is set to 0.

This if is executed if the exponent is 0 but the argument is not 0, i.e. is "too large".

please see:
https://github.com/php/php-src/blob/51d93c194720b2ec0945236667eb9dd0a91d64a2/ext/bcmath/libbcmath/src/num2long.c#L54

Reply:
Added a comment in https://github.com/php/php-src/pull/13741/commits/bb17dafb69c70839c44f4d9d53bcf44d48a8d677